core/keyboard_nav/key_map.js

```diff
@@ -0,0 +1,250 @@
+  if (key) {
+    delete Blockly.user.keyMap.map_[key];
+  }
+  Blockly.user.keyMap.map_[keyCode] = action;
+  // If the action is not in the actions array add it.
+  if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+    Blockly.user.keyMap.actions_.push(action);
+  }
+};
+
+/**
+ * Creates a new key map and a new actions array.
+ * @param {Object<string, Blockly.KeyAction>} keyMap The object holding the key
+ *     to action mapping.
+ * @package
+ */
+Blockly.user.keyMap.setKeyMap = function(keyMap) {
+  var keys = Object.keys(keyMap);
+  // Update the actions array to include any new actions found in the new key map.
+  for (var i = 0; i < keys.length; i++) {
+    var action = keyMap[keys[i]];
+    if (Blockly.user.keyMap.actions_.indexOf(action) === -1) {
+      Blockly.user.keyMap.actions_.push(action);
+    }
+  }
+  Blockly.user.keyMap.map_ = keyMap;
+};
+
+/**
+ * Gets the current key map.
+ * @return {Object<string,Blockly.KeyAction>} The object holding the key to
+ *     action mapping.
+ * @package
+ */
+Blockly.user.keyMap.getKeyMap = function() {
+  return Blockly.user.keyMap.map_;
+};
+
+/**
+ * Get the action by the serialized key code.
+ * @param {string} keyCode The serialized key code.
+ * @return {Blockly.KeyAction|undefined} The action holding the function to
+ *     call when the given keyCode is used or undefined if no action exists.
+ * @package
+ */
+Blockly.user.keyMap.getActionByKeyCode = function(keyCode) {
+  return Blockly.user.keyMap.map_[keyCode];
+};
+
+/**
+ * Get the serialized key that corresponds to the action name.
+ * @param {!string} actionName The name of the action.
+ * @return {string} The serialized key or null if the action does not have
+ *     a key mapping.
+ * @package
+ */
+Blockly.user.keyMap.getKeyByActionName = function(actionName) {
+  var keys = Object.keys(Blockly.user.keyMap.map_);
+  for (var i = 0; i < keys.length; i++) {
+    var key = keys[i];
+    var action = Blockly.user.keyMap.map_[key];
+    if (actionName === action.name) {
+      return key;
+    }
+  }
+  return null;
+};
+
+/**
+ * Serialize the key event.
+ * @param {!Event} e A key up event holding the key code.
+ * @return {!string} A string containing the serialized key event.
+ */
+Blockly.user.keyMap.serializeKeyEvent = function(e) {
+  var key = "";
+  if (e.getModifierState('Shift')) {
+    key += '+';
+  }
+  if (e.getModifierState('Control')) {
+    key += 'ctrl';
+  }
+  if (e.getModifierState('Alt')) {
+    key += 'alt';
+  }
+  if (e.getModifierState('Meta')) {
+    key += 'meta';
+  }
+  key += e.keyCode;
+  return key;
+};
+
+/**
+ * Creates the default key map.
+ * @return {!Object<string,Blockly.KeyAction>} An object holding the default key
+ *     to action mapping.
+ */
+Blockly.user.keyMap.createDefaultKeyMap = function() {
+  //Set W
+  var wAction = new Blockly.KeyAction('previous', 'Goes to the previous location', function() {
```

Comment:
These actions should be public constants on the navigation namespace called things like "ACTION_PREVIOUS". That way, developers can call setActionForKey(keycode, Blockly.navigation.ACTION_PREVIOUS).

Reply:
done